### PR TITLE
Swap `forEachChild` to use a table of functions instead of a `switch` statement.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
`forEachChild` is one of the core pieces of machinery we have for traversing nodes in our trees. It's coded using a fairly large `switch` statement; however, while reading up on an emulator built in JavaScript (https://artemis.sh/2022/08/07/emulating-calculators-fast-in-js.html), the author called out that most engines *do not* seem to do the jump-table optimization that some compilers use for `switch` statements.

So this change introduces a function table for all of `forEachChild`. It seems to cut off around 50ms - 150ms from each perf test's compile time, and speeds up several operations in the language service.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this faster

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this faster

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test this

**Review by sheetalkamat — APPROVED** [review]

**Comment by rbuckton** [comment]
I wonder if the same benefit could be found in `visitEachChild` as well, or potentially any of the main `visitor` functions in each transformer. This does create an array with quite a bit of empty space due to how many tokens don't have or need a visitor. I wonder if an object literal would have better or worse performance metrics, i.e.:

```ts
const forEachChildTable: Partial<Record<SyntaxKind, ForEachChildFunction>> = {
  [SyntaxKind.QualifiedName]: forEachQualifiedName,
  [SyntaxKind.TypeParameter]: forEachTypeParameter,
  // etc.
};
```

**Comment by DanielRosenwasser** [comment]
> I wonder if an object literal would have better or worse performance metrics,

If an object has identical perf, then I'd assume it's an indication that the array being constructed is sparse - but we need the look-up to be fast. If there's an easy way to create a definitely-packed array, then I'll switch to it here.

All the tests do admittedly indicate higher memory usage - but more than I would've expected.

> I wonder if the same benefit could be found in `visitEachChild` as well

It's definitely worth playing with, especially since `visitEachChild` tends to be pretty much identical to `forEachChild`. I don't currently see a difference with `checkExpressionWorker` though, but I'm playing with it over at https://github.com/microsoft/TypeScript/pull/50228.

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -95,6 +95,979 @@ namespace ts {
         return isMetaProperty(node) && node.keywordToken === SyntaxKind.ImportKeyword && node.name.escapedText === "meta";
     }
 
+    type ForEachChildFunction = <T>(node: any, cbNode: (node: Node) => T | undefined, cbNodes?: (nodes: NodeArray<Node>) => T | undefined) => T | undefined;
+
+    const forEachChildTable = new Array<ForEachChildFunction>(SyntaxKind.Count);
```
Comment (amcasey) [inline src/compiler/parser.ts:100]: Some thoughts on this initialization mechanism:
1. AFAIK, the ctor arg does nothing (@jakebailey?)
2. The array will always be sparse (among other things, we're always skipping the first 160 elements)
3. We don't have any protection against specifying two different functions for the same kind.

**Review by amcasey — APPROVED** [review]
Binding isn't very important as a percentage of tsc execution time, but this should help in editor scenarios where we rebind regularly.

**Comment by amcasey** [comment]
Is this something we could move to its own file?

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
The latest change swaps to `push`ing elements as suggested in https://v8.dev/blog/elements-kinds#avoid-creating-holes. I might try this on the other PRs (#50228, #50238), but it seems like those do not get any wins from the table-based approach.

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -95,6 +95,980 @@ namespace ts {
         return isMetaProperty(node) && node.keywordToken === SyntaxKind.ImportKeyword && node.name.escapedText === "meta";
     }
 
+    type ForEachChildFunction = <T>(node: any, cbNode: (node: Node) => T | undefined, cbNodes?: (nodes: NodeArray<Node>) => T | undefined) => T | undefined;
+
+    const forEachChildTable: Array<ForEachChildFunction | undefined> = [];
+    for (let i = 0; i < SyntaxKind.Count; i++) forEachChildTable.push(undefined);
```
Comment (amcasey) [inline src/compiler/parser.ts:101]: As long as we're micro-optimizing, I wonder if it would be happier with a no-op function.

**Review by weswigham — COMMENTED** [review]
As a minor semantic nit, can we rename all the new sub-functions along the lines of `forEachInNodeKind` rather than `forEachNodeKind` - the former sounds more like iterating the children (which this does), rather than filtering to only the nodes of the given kind within some arbitrary node (which this does not) - specifically because we already have functions like `forEachReturnStatement` which do the later.

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this faster

**Comment by DanielRosenwasser** [comment]
I just realized that our down-level code for

```ts
forEachChildTable[node.kind]?.(node, cbNode, cbNodes);
```

is

```js
return (_a = forEachChildTable[node.kind]) === null || _a === void 0 ? void 0 : _a.call(forEachChildTable, node, cbNode, cbNodes);
```

So we have to

* check for `null` in addition to `undefined`
* pass in the target object every single time
* run `.call`

I'm going to see how the no-op function is, but I'm also going to try to rewrite the code so that we create a variable for the function and call that directly.

**Comment by rbuckton** [comment]
Why are you bothering to call the noop in the first place? Any invocation has overhead, even if its potentially inlined. Have you tried anything like the following to avoid the invocation when it's unnecessary?

```js
const fn = forEachChildTable[node.kind];
return fn === forEachChildNoOpFn ? undefined : fn(node, cbNode, cbNodes);
```

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

File: src/compiler/parser.ts (hunk 2 above)
Comment (DanielRosenwasser) [reply src/compiler/parser.ts:101]: There might've been something I missed, but it didn't seem to be.

Comment (amcasey) [reply src/compiler/parser.ts:101]: Interesting.  Maybe `undefined` is already in the domain.  Thanks for checking!

**Comment by DanielRosenwasser** [comment]
I like the current implementation. I'm open to moving it to another file or something if we want.

File: src/compiler/parser.ts (hunk 1 above)
Comment (jakebailey) [reply src/compiler/parser.ts]: Yes, AFAIK the size argument doesn't actually make this a compact array or anything, hence them trying to add a `Array.withCapacity` instead or something.

**Comment by jakebailey** [comment]
My main concern with declaring functions for all of these is that someone might actually call them when they shouldn't; I think it might be worth sticking all of these into a namespace whose only export is the array; that way you can't actually go use any of the functions directly, but the contents of the array will still be identical because they are still local to the array.

**Review by weswigham — APPROVED** [review]
I'm reasonably sure why this is more performant, and I'm reasonably sure it's not an always-correct kinda thing like the linked issued involving `switch` implies after trying #50245 (and has more to do with the (instruction-length-wise) size of the function bodies involved and how harry they become to optimize when they're larger). Given that, `visitEachChild` in `visitorPublic.ts` is pretty close in appearance to this function and would probably benefit from the smaller size in much the same way - likely moreso than `checkSourceElementWorker` looks to over at #50238.

**Review by rbuckton — APPROVED** [review]
I don't have a strong preference on whether we use an array or object literal for this given they have similar perf results. Given how many functions were added for this, I do think it would make sense to split this out into a separate file at some point.

**Comment by rbuckton** [comment]
I created #50266 to do the same kind of thing for `visitEachChild` but used the object literal route instead.

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -95,6 +95,980 @@ namespace ts {
...
+    }
+    function forEachChildInJSDocLinkCodeOrPlain<T>(node: JSDocLink | JSDocLinkCode | JSDocLinkPlain, cbNode: (node: Node) => T | undefined, _cbNodes?: (nodes: NodeArray<Node>) => T | undefined): T | undefined {
+            return visitNode(cbNode, node.name);
+    }
+
+    function forEachChildInJSDocTypeLiteral<T>(node: JSDocTypeLiteral, cbNode: (node: Node) => T | undefined, _cbNodes?: (nodes: NodeArray<Node>) => T | undefined): T | undefined {
+        return forEach(node.jsDocPropertyTags, cbNode);
+    }
+
+
```
Comment (jakebailey) [inline src/compiler/parser.ts:1062]: Silly nit: spacing got a little wonky here.

**Review by jakebailey — APPROVED** [review]
Really nice perf find; I do wish that this were more type safe (I kind of prefer the object version in #50266, but I'm not sure that same thing can apply here as that uses a huge union which I think we don't want for all `Node`).

I would eventually like to see it in another file to help hide all of this from other functions in the parser (I have a little bit of PTSD after the previous visitor-wrong-function-bugs).

**Comment by sandersn** [comment]
@DanielRosenwasser piles of people have signed off on this PR; should we merge it for 4.9?

**Comment by DanielRosenwasser** [comment]
It sounds like Ron and Jake are both on board with introducing an internal `HasChildren` type, and then using a sparse object literal to enforce that everything is set correctly. I think that's a good idea, and we could always swap around the implementations if we don't like these.

So I think we should get one more sign-off on Ron's change first at #50266. Once that's in, I'll update this PR to leverage the same `HasChildren` trick.

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this faster

**Comment by DanielRosenwasser** [comment]
Because integration with the `HasChildren` union is such a pain - because `forEachChild` doesn't account for `Identifier`s with children, and because `visitEachChild` doesn't account for any JSDoc nodes - I'm going to defer it to a different PR.

I want to be able to untangle the perf wins here with whatever new stuff that might introduce.

File: src/compiler/parser.ts [hunk 4]
```diff
@@ -95,6 +95,737 @@ namespace ts {
...
+    }
+
+    function forEachChildInObjectOrArrayBindingPattern<T>(node: BindingPattern, cbNode: (node: Node) => T | undefined, cbNodes?: (nodes: NodeArray<Node>) => T | undefined): T | undefined {
+        return visitNodes(cbNode, cbNodes, node.elements);
+    }
+
+    function forEachChildInCallOrNewExpression<T>(node: CallExpression | NewExpression, cbNode: (node: Node) => T | undefined, cbNodes?: (nodes: NodeArray<Node>) => T | undefined): T | undefined {
+        return visitNode(cbNode, node.expression) ||
+            // TODO: should we separate these branches out?
+            visitNode(cbNode, (node as CallExpression).questionDotToken) ||
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts:759]: I'll do this for a follow-up change.

**Comment by jakebailey** [comment]
Those perf results make me happy every time I see them.

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -95,6 +95,737 @@ namespace ts {
...
+        [SyntaxKind.JSDocTypedefTag]: function forEachChildInJSDocTypedefTag<T>(node: JSDocTypedefTag, cbNode: (node: Node) => T | undefined, cbNodes?: (nodes: NodeArray<Node>) => T | undefined): T | undefined {
+            return visitNode(cbNode, node.tagName) ||
+                (node.typeExpression &&
+                    node.typeExpression.kind === SyntaxKind.JSDocTypeExpression
+                    ? visitNode(cbNode, node.typeExpression) ||
+                    visitNode(cbNode, node.fullName) ||
+                    (typeof node.comment === "string" ? undefined : visitNodes(cbNode, cbNodes, node.comment))
+                    : visitNode(cbNode, node.fullName) ||
+                    visitNode(cbNode, node.typeExpression) ||
+                    (typeof node.comment === "string" ? undefined : visitNodes(cbNode, cbNodes, node.comment)));
```
Comment (rbuckton) [inline src/compiler/parser.ts:703]: The indentation here makes it a bit difficult to read.

**Review by rbuckton — APPROVED** [review]

File: src/compiler/parser.ts (hunk 5 above)
Comment (DanielRosenwasser) [reply src/compiler/parser.ts:703]: I agree, it's awful - but I believe it's from the original code. I'll fix it up in the `HasChildren` PR.

Comment (DanielRosenwasser) [reply src/compiler/parser.ts:703]: I guess I have some lints I have to fix so... I guess I'll fix now.

**Comment by leidegre** [comment]
This was a welcome addition. Nice work @DanielRosenwasser 

I'm just going to leave this here as something I've noticed. About enums, switches and performance in V8.

V8 doesn't like switches with enums, it does however deal with const enums very differently. I wonder what would happen if the SyntaxKind enum was made into a const enum (and you kept the switch). Changing the SyntaxKind enum into a cost enum creates other problems but I think the performance "problem" with the switch statement has more to do so that it isn't being generated into a jump table in machine code when you don't use literal integer numbers (SMIs) as case labels.

I don't remember the exact number but I had a switch with 250+ cases in it and when I changed from enum to const enum it made a huge difference.

**Comment by rbuckton** [comment]
> [...] I wonder what would happen if the SyntaxKind enum was made into a const enum (and you kept the switch). 

`SyntaxKind` is defined as a `const enum` internally, so it was using literal integers for case labels. It's likely we were running afoul of some other issue that resulted in a suboptimal compiled representation.

**Comment by leidegre** [comment]
> > [...] I wonder what would happen if the SyntaxKind enum was made into a const enum (and you kept the switch).
> 
> `SyntaxKind` is defined as a `const enum` internally, so it was using literal integers for case labels. It's likely we were running afoul of some other issue that resulted in a suboptimal compiled representation.

Oh really, interesting! Thanks for letting me know.

**Pull request merged**